django/db/models/fields/tuple_lookups.py

```diff
@@ -31,29 +31,22 @@ def check_tuple_lookup(self):
     def check_rhs_is_tuple_or_list(self):
         if not isinstance(self.rhs, (tuple, list)):
             raise ValueError(
-                f"'{self.lookup_name}' lookup of '{self.lhs.field.name}' field "
-                "must be a tuple or a list"
```

Comment:
This was actually fine :) as it doesn't have any method/function calls.

Reply:
... and this is still untested, do we need this check?

Reply:
Please view this as a sanity check for now. There's no harm in having it, it adds some extra safety. And it will be needed directly with the upcoming [composite primary keys patch](https://github.com/django/django/pull/18056/files#diff-e2030abd6e95d46826f9e813df5e58c195dce4f83241d046229cfddfe6bb2238R97).

Reply:
> ... it adds some extra safety

How? since it's unreachable :thinking: Do you mean for someone how will override our lookups and will try to do the same? 

> There's no harm in having it...

Pretty sure someone will send PR in the future to remove it as a dead code.

Reply:
I think it's a useful sanity check to make, the same way we make assertions all throughout the codebase, it's useful to check the type here too, since these lookups only support direct values in the form of a `tuple` or a `list`. It's similar to `assert isinstance(self.lhs, ColPairs)` above, which you don't seem to have a problem with although this `AssertionError` is unreachable. Maybe you would prefer `AssertionError` over `ValueError`?

Reply:
We can remove it here and have it added as part of #18056 which has tests to cover it

Reply:
@felixxm ?

Reply:
@sarahboyce I see you went with removing the `assert` as well? So does that mean we're not supposed to add `assert`s if we're asserting something that's always true? Let me know and I'll keep it in mind next time.

Reply:
> So does that mean we shouldn't add asserts if we're asserting something that's always true?

I think we shouldn't - similar to how we wouldn't add a if condition for something always true

Reply:
> I think we shouldn't - similar to how we wouldn't add a if condition for something always true

Hmm, is that not the point of `assert` though? So wouldn't that imply we shouldn't use `assert` at all? It doesn't really matter in this case as I was planning on making these classes a bit more versatile anyway. It's more of a philosophical question.

Reply:
I don't mean to imply that. To me, it's like doing
```python
if not isinstance(self.lhs, ColPairs):
    raise ...
```
If users can trigger the error, then great we are controlling the error message to something hopefully more useful, and we can write a test to cover this (so `assert` would be fine)

Reply:
FYI, here are the changes I mentioned I wanted to make: https://github.com/django/django/pull/18572